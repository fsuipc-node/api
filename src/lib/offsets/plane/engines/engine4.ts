import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine4: OffsetList = {
  engine4ThrottleLever: new Offset({
    value: 0xA54,
    name: 'engine4ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine4PropLever: new Offset({
    value: 0xA56,
    name: 'engine4PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4MixtureLever: new Offset({
    value: 0xA58,
    name: 'engine4MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4StarterSwitchPosition: new Offset({
    value: 0xA5A,
    name: 'engine4StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine4 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine4Firing: new Offset({
    value: 0xA5C,
    name: 'engine4Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine4N2: new Offset({
    value: 0xA5E,
    name: 'engine4N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4N1: new Offset({
    value: 0xA60,
    name: 'engine4N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4PropRPM: new Offset({
    value: 0xA60,
    name: 'engine4PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4ThrottleLeverControl: new Offset({
    value: 0xA62,
    name: 'engine4ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine4 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine4FuelFlowLbHourSSL: new Offset({
    value: 0xA68,
    name: 'engine4FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine4 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine4AntiIce: new Offset({
    value: 0xA7A,
    name: 'engine4AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine4 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine4OilTemp: new Offset({
    value: 0xA80,
    name: 'engine4OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine4 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine4OilPres: new Offset({
    value: 0xA82,
    name: 'engine4OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine4 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine4PressureRatio: new Offset({
    value: 0xA84,
    name: 'engine4PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine4 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.Int16,
    permission: 'r',
  }),
  engine4EGT: new Offset({
    value: 0xA86,
    name: 'engine4EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine4 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine4MP: new Offset({
    value: 0xA88,
    name: 'engine4MP',
    category: OffsetCategory.ENGINE,
    description: 'engine4 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine4RPMScaler: new Offset({
    value: 0xA90,
    name: 'engine4RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine4 RPM Scaler - props: used to calculate RPM',
    type: Type.Int16,
    permission: 'r',
  }),
  engine4OilQuantity: new Offset({
    value: 0xA98,
    name: 'engine4OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine4 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4Vibration: new Offset({
    value: 0xA9C,
    name: 'engine4Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine4 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4HydPres: new Offset({
    value: 0xAA0,
    name: 'engine4HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine4 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4HydQuantity: new Offset({
    value: 0xAA4,
    name: 'engine4HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine4 hydraulic quantity - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4CHT: new Offset({
    value: 0xAB0,
    name: 'engine4CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine4 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine4ITT: new Offset({
    value: 0xAB8,
    name: 'engine4ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine4 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4Torque: new Offset({
    value: 0xABC,
    name: 'engine4Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine4 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4FuelPres: new Offset({
    value: 0xAC0,
    name: 'engine4FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine4 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.Int32,
    permission: 'r',
  }),
  engine4FuelUsedSinceStart: new Offset({
    value: 0xAD4,
    name: 'engine4FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine4 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine4FuelElapsedTime: new Offset({
    value: 0xAD8,
    name: 'engine4FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine4 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine4FuelFlowLbHour: new Offset({
    value: 0xAE0,
    name: 'engine4FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine4 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TorqueFtLb: new Offset({
    value: 0xAE8,
    name: 'engine4TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine4 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine4TurbineN1: new Offset({
    value: 0x2300,
    name: 'engine4TurbineN1',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine N1 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineN2: new Offset({
    value: 0x2308,
    name: 'engine4TurbineN2',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine N2 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineCorrectedN1: new Offset({
    value: 0x2310,
    name: 'engine4TurbineCorrectedN1',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine corrected N1 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineCorrectedN2: new Offset({
    value: 0x2318,
    name: 'engine4TurbineCorrectedN2',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine corrected N2 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineCorrectedFuelFlow: new Offset({
    value: 0x2320,
    name: 'engine4TurbineCorrectedFuelFlow',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine corrected fuel flow - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineTorque: new Offset({
    value: 0x2328,
    name: 'engine4TurbineTorque',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine corrected fuel flow - 0.0 to 1.0',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineEPR: new Offset({
    value: 0x2330,
    name: 'engine4TurbineEPR',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine EPR',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineITT: new Offset({
    value: 0x2338,
    name: 'engine4TurbineITT',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine ITT',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineAfterburnerActive: new Offset({
    value: 0x2348,
    name: 'engine4TurbineAfterburnerActive',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine afterburner active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine4TurbineJetThrust: new Offset({
    value: 0x234C,
    name: 'engine4TurbineJetThrust',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine jet thrust',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineTankSelector: new Offset({
    value: 0x2354,
    name: 'engine4TurbineTankSelector',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine tank selector',
    convert: 'fuelTank',
    mapping: true,
    type: Type.UInt32,
    permission: 'r',
  }),
  engine4TurbineUsedTank: new Offset({
    value: 0x2358,
    name: 'engine4TurbineUsedTank',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine used tank',
    type: Type.BitArray,
    length: 11,
    permission: 'r',
  }),
  engine4TurbineAvailableTankCount: new Offset({
    value: 0x235C,
    name: 'engine4TurbineAvailableTankCount',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine available tank count',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine4TurbineFuelFlow: new Offset({
    value: 0x2360,
    name: 'engine4TurbineFuelFlow',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine fuel flow',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineIsFuelAvailable: new Offset({
    value: 0x2368,
    name: 'engine4TurbineIsFuelAvailable',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine is fuel available flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine4TurbineBleedPressure: new Offset({
    value: 0x236C,
    name: 'engine4TurbineBleedPressure',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine bleed air pressure - lb/sq in',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineReverser: new Offset({
    value: 0x237C,
    name: 'engine4TurbineReverser',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine reverser position - 0.0 - 1.0 lb/sq in',
    convert: 'Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineVibration: new Offset({
    value: 0x2384,
    name: 'engine4TurbineVibration',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine vibration',
    type: Type.Double,
    permission: 'r',
  }),
  engine4TurbineIgnition: new Offset({
    value: 0x238C,
    name: 'engine4TurbineIgnition',
    category: OffsetCategory.ENGINE,
    description: 'engine 4 turbine is fuel available flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
};
