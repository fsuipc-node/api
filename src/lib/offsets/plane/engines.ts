import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

import { engine1 } from './engines/engine1';
import { engine2 } from './engines/engine2';
import { engine3 } from './engines/engine3';
import { engine4 } from './engines/engine4';

export const engines: OffsetList = {
  engineType: new Offset({
    value: 0x609,
    name: 'engineType',
    category: OffsetCategory.ENGINE,
    description: 'engine type',
    convert: 'engineType',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  hasMixtureControl: new Offset({
    value: 0x780,
    name: 'hasMixtureControl',
    category: OffsetCategory.ENGINE,
    description: 'has mixture control',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.ENGINE,
    description: 'has carb heat',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  activeEngine: new Offset({
    value: 0x888,
    name: 'activeEngine',
    category: OffsetCategory.ENGINE,
    description: 'active engine pattern',
    type: Type.BitArray,
    length: 1,
    permission: 'rw',
  }),
  ...engine1,
  ...engine2,
  ...engine3,
  ...engine4,
};
