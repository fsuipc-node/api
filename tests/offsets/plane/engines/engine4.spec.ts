import { engine4 as offsets } from '@offsets/plane/engines/engine4';

describe('offset - plane/engines/engine4', () => {
  const offsetsTestCases = [
    { name: 'engine4ThrottleLever', value: -2048, expectedResult: -50 },
    { name: 'engine4ThrottleLever', value: 0, expectedResult: 0 },
    { name: 'engine4ThrottleLever', value: 8192, expectedResult: 50 },
    { name: 'engine4PropLever', value: -2048, expectedResult: -50 },
    { name: 'engine4PropLever', value: 0, expectedResult: 0 },
    { name: 'engine4PropLever', value: 8192, expectedResult: 50 },
    { name: 'engine4MixtureLever', value: 8192, expectedResult: 50 },
    { name: 'engine4Firing', value: 0, expectedResult: false },
    { name: 'engine4Firing', value: 1, expectedResult: true },
    { name: 'engine4N2', value: 8192, expectedResult: 50 },
    { name: 'engine4N1', value: 8192, expectedResult: 50 },
    { name: 'engine4PropRPM', value: 107374182400, expectedResult: 100 },
    { name: 'engine4FuelFlowLbHourSSL', value: 256, expectedResult: 2 },
    { name: 'engine4AntiIce', value: 0, expectedResult: false },
    { name: 'engine4AntiIce', value: 1, expectedResult: true },
    { name: 'engine4OilTemp', value: 10532, expectedResult: 90 },
    { name: 'engine4OilPres', value: 22342, expectedResult: 75 },
    { name: 'engine4PressureRatio', value: 20480, expectedResult: 2 },
    { name: 'engine4EGT', value: 14281, expectedResult: 750 },
    { name: 'engine4MP', value: 20480, expectedResult: 20 },
    { name: 'engine4OilQuantity', value: 8167, expectedResult: 50 },
    { name: 'engine4Vibration', value: 6553, expectedResult: 2 },
    { name: 'engine4HydPres', value: 64, expectedResult: 16 },
    { name: 'engine4HydQuantity', value: 8192, expectedResult: 50 },
    { name: 'engine4ITT', value: 3276800, expectedResult: 200 },
    { name: 'engine4Torque', value: 8192, expectedResult: 50 },
    { name: 'engine4FuelPres', value: 288, expectedResult: 2 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = offsets[testedOffset.name].convert.replace(new RegExp(/{VAL}/g), testedOffset.value.toString());

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
