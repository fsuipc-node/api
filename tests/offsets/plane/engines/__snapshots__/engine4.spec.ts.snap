// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/engines/engine4 offsets list should have required properties 1`] = `
Object {
  "engine4AntiIce": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine4 anti-ice/carb heat active",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4AntiIce",
    "permission": "r",
    "type": 5,
    "value": 2682,
  },
  "engine4CHT": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 cylinder head temperature - F",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4CHT",
    "permission": "r",
    "type": 8,
    "value": 2736,
  },
  "engine4EGT": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 860 / 16384)",
    "description": "engine4 exhaust gas temperature - Trust only on jet engine - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4EGT",
    "permission": "r",
    "type": 2,
    "value": 2694,
  },
  "engine4Firing": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 4 combustion",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4Firing",
    "permission": "r",
    "type": 5,
    "value": 2652,
  },
  "engine4FuelElapsedTime": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 fuel elapsed time - h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4FuelElapsedTime",
    "permission": "r",
    "type": 9,
    "value": 2776,
  },
  "engine4FuelFlowLbHour": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 fuel elapsed time - lb/h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4FuelFlowLbHour",
    "permission": "r",
    "type": 8,
    "value": 2784,
  },
  "engine4FuelFlowLbHourSSL": Offset {
    "category": "engine",
    "convert": "{VAL} / 128",
    "description": "engine4 fuel flow (lb per hour, sea level)",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4FuelFlowLbHourSSL",
    "permission": "r",
    "type": 5,
    "value": 2664,
  },
  "engine4FuelPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 144",
    "description": "engine4 fuel pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4FuelPres",
    "permission": "r",
    "type": 3,
    "value": 2752,
  },
  "engine4FuelUsedSinceStart": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 used fuel since start - lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4FuelUsedSinceStart",
    "permission": "r",
    "type": 9,
    "value": 2772,
  },
  "engine4HydPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 4",
    "description": "engine4 hydraulic pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4HydPres",
    "permission": "r",
    "type": 3,
    "value": 2720,
  },
  "engine4HydQuantity": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine4 hydraulic quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4HydQuantity",
    "permission": "r",
    "type": 3,
    "value": 2724,
  },
  "engine4ITT": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384",
    "description": "engine4 turbine temperature - C",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4ITT",
    "permission": "r",
    "type": 3,
    "value": 2744,
  },
  "engine4MP": Offset {
    "category": "engine",
    "convert": "{VAL} / 1024",
    "description": "engine4 manifold pressure - inHg",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4MP",
    "permission": "r",
    "type": 5,
    "value": 2696,
  },
  "engine4MixtureLever": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 4 prop lever - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4MixtureLever",
    "permission": "rw",
    "type": 2,
    "value": 2648,
  },
  "engine4N1": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 4 N1 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4N1",
    "permission": "rw",
    "type": 2,
    "value": 2656,
  },
  "engine4N2": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 4 N2 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4N2",
    "permission": "rw",
    "type": 2,
    "value": 2654,
  },
  "engine4OilPres": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 55 / 16384)",
    "description": "engine4 oil pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4OilPres",
    "permission": "r",
    "type": 5,
    "value": 2690,
  },
  "engine4OilQuantity": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} / 16384 * 100)",
    "description": "engine4 oil quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4OilQuantity",
    "permission": "r",
    "type": 3,
    "value": 2712,
  },
  "engine4OilTemp": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 140 / 16384)",
    "description": "engine4 oil temp - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4OilTemp",
    "permission": "r",
    "type": 2,
    "value": 2688,
  },
  "engine4PressureRatio": Offset {
    "category": "engine",
    "convert": "{VAL} * 1.6 / 16384",
    "description": "engine4 pressure ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4PressureRatio",
    "permission": "r",
    "type": 2,
    "value": 2692,
  },
  "engine4PropLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 4 prop lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4PropLever",
    "permission": "rw",
    "type": 2,
    "value": 2646,
  },
  "engine4PropRPM": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 / 65536",
    "description": "derive RPM by multiplying by  - negative = counter-rotating propeller",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4PropRPM",
    "permission": "rw",
    "type": 2,
    "value": 2656,
  },
  "engine4RPMScaler": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 RPM Scaler - props: used to calculate RPM",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4RPMScaler",
    "permission": "r",
    "type": 2,
    "value": 2704,
  },
  "engine4StarterSwitchPosition": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4StarterSwitchPosition",
    "permission": "rw",
    "type": 5,
    "value": 2650,
  },
  "engine4ThrottleLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 4 throttle lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4ThrottleLever",
    "permission": "r",
    "type": 2,
    "value": 2644,
  },
  "engine4ThrottleLeverControl": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 throttle lever control - -4096 to +16384",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4ThrottleLeverControl",
    "permission": "rw",
    "type": 2,
    "value": 2658,
  },
  "engine4Torque": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine4 torque - trust only for helo - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4Torque",
    "permission": "r",
    "type": 3,
    "value": 2748,
  },
  "engine4TorqueFtLb": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine4 fuel elapsed time - ft lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TorqueFtLb",
    "permission": "r",
    "type": 9,
    "value": 2792,
  },
  "engine4TurbineAfterburnerActive": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 4 turbine afterburner active",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineAfterburnerActive",
    "permission": "r",
    "type": 6,
    "value": 9032,
  },
  "engine4TurbineAvailableTankCount": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine available tank count",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineAvailableTankCount",
    "permission": "r",
    "type": 6,
    "value": 9052,
  },
  "engine4TurbineBleedPressure": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine bleed air pressure - lb/sq in",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineBleedPressure",
    "permission": "r",
    "type": 8,
    "value": 9068,
  },
  "engine4TurbineCorrectedFuelFlow": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine corrected fuel flow - lb/h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineCorrectedFuelFlow",
    "permission": "r",
    "type": 8,
    "value": 8992,
  },
  "engine4TurbineCorrectedN1": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine corrected N1 - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineCorrectedN1",
    "permission": "r",
    "type": 8,
    "value": 8976,
  },
  "engine4TurbineCorrectedN2": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine corrected N2 - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineCorrectedN2",
    "permission": "r",
    "type": 8,
    "value": 8984,
  },
  "engine4TurbineEPR": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine EPR",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineEPR",
    "permission": "r",
    "type": 8,
    "value": 9008,
  },
  "engine4TurbineFuelFlow": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine fuel flow",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineFuelFlow",
    "permission": "r",
    "type": 8,
    "value": 9056,
  },
  "engine4TurbineITT": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine ITT",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineITT",
    "permission": "r",
    "type": 8,
    "value": 9016,
  },
  "engine4TurbineIgnition": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 4 turbine is fuel available flag",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineIgnition",
    "permission": "rw",
    "type": 6,
    "value": 9100,
  },
  "engine4TurbineIsFuelAvailable": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 4 turbine is fuel available flag",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineIsFuelAvailable",
    "permission": "r",
    "type": 6,
    "value": 9064,
  },
  "engine4TurbineJetThrust": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine jet thrust",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineJetThrust",
    "permission": "r",
    "type": 8,
    "value": 9036,
  },
  "engine4TurbineN1": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine N1 - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineN1",
    "permission": "r",
    "type": 8,
    "value": 8960,
  },
  "engine4TurbineN2": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine N2 - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineN2",
    "permission": "r",
    "type": 8,
    "value": 8968,
  },
  "engine4TurbineReverser": Offset {
    "category": "engine",
    "convert": "Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))",
    "description": "engine 4 turbine reverser position - 0.0 - 1.0 lb/sq in",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineReverser",
    "permission": "r",
    "type": 8,
    "value": 9084,
  },
  "engine4TurbineTankSelector": Offset {
    "category": "engine",
    "convert": "fuelTank",
    "description": "engine 4 turbine tank selector",
    "length": undefined,
    "mapping": true,
    "name": "engine4TurbineTankSelector",
    "permission": "r",
    "type": 6,
    "value": 9044,
  },
  "engine4TurbineTorque": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine corrected fuel flow - 0.0 to 1.0",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineTorque",
    "permission": "r",
    "type": 8,
    "value": 9000,
  },
  "engine4TurbineUsedTank": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine used tank",
    "length": 11,
    "mapping": undefined,
    "name": "engine4TurbineUsedTank",
    "permission": "r",
    "type": 12,
    "value": 9048,
  },
  "engine4TurbineVibration": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine 4 turbine vibration",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4TurbineVibration",
    "permission": "r",
    "type": 8,
    "value": 9092,
  },
  "engine4Vibration": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 5 / 16384)",
    "description": "engine4 vibration",
    "length": undefined,
    "mapping": undefined,
    "name": "engine4Vibration",
    "permission": "r",
    "type": 3,
    "value": 2716,
  },
}
`;
